Implemented messaging translation

garzeah opened the pull request:
**Updates**
- Created a new user route to obtain a list a languages for translation
- Created an input to select a language upon registration
- Updated user model to account for languages
- Implemented a toggle to see translated/non-translated messages
- Google translate is now working
      - How it works is that every message that a user receives will be translated to their designated language they chose upon sign up.

**Future**
- Creating a modal for users to edit their profile
- Creating a demo button

**Questions**
- I know we aren't supposed to copy the original completely but do you think there are any features I should add?

**Pictures**
**Translated**
<img width="858" alt="translated" src="https://user-images.githubusercontent.com/29414787/109918884-9602a500-7c6c-11eb-9a57-907dc916d044.png">

**Original**
<img width="858" alt="original" src="https://user-images.githubusercontent.com/29414787/109918907-a024a380-7c6c-11eb-9961-d0a5830eec41.png">

garzeah committed:
testing production website

garzeah committed:
removed config folder and migrated everything to env

garzeah committed:
testing prod website

garzeah committed:
implemented google translate

garzeah committed:
refactored code so google api works on prod server

garzeah committed:
changed recipient message color

garzeah committed:
updated readme

garzeah committed:
fixed typo on readme

garzeah committed:
made login the default page

bonnieli commented on controllers/convoController.js:
```diff
@@ -104,10 +108,36 @@ const fetchAllMyConversationsGet = async (req, res) => {
 
 // Get a specific conversation and its messages
 const fetchConversationGet = async (req, res) => {
+	const { id, lang } = req.query;
```
since language is part of the user model, which is loaded through your middleware, you can actually access the user's language without having to pass it into your query! you can probably call `req.user.language` here

garzeah replied:
Thanks just made the change! I updated my middleware to have the user information, changed the query method, and controller.

bonnieli commented on controllers/convoController.js:
```diff
@@ -104,10 +108,36 @@ const fetchAllMyConversationsGet = async (req, res) => {
 		// Finding all messages associated with a specific conversation
-		const conversation = await Message.find({ conversationID: req.params.id });
-		res.send(conversation);
+		let conversation = await Message.find({ conversationID: id });
```
ah we may want to rethink our Message model because instead of translating the same content each time you update or grab your conversation, we want a more efficient way of doing this - or else you will reach your rate limit for your google api very soon!

bonnieli replied:
we can talk about this tomorrow in our call!

garzeah replied:
> ah we may want to rethink our Message model because instead of translating the same content each time you update or grab your conversation, we want a more efficient way of doing this - or else you will reach your rate limit for your google api very soon!

Oh yeah huh, that's a good point. Looking forward to talking more about it tomorrow!

bonnieli replied:
at most translate each message once per message per language

garzeah committed:
updated middleware, convoController, and query method

garzeah committed:
optimized google-translate so a message gets translated once at most per language

garzeah committed:
added comments

garzeah commented:
Hey Bonnie, just optimized the google translate part so that each message gets translated once per language. Going to start working on redux tomorrow!

**Updates**
- Updated Message.js model
- Cleaned some code
- Optimized Google Translate logic
- Updated front end to reflect back end changes

I was wondering what's the best way to deal with collections w/ documents that have nested arrays? For example, in my convoController.js file in my fetchAllMyConversationsGet function I have to iterate through every conversation and in each conversation I have to iterate through each participant just to find user ID's that are not mine. The way I wrote the code is really inefficient and could be improved a lot.

bonnieli reviewed (approved):
great revision! :)

bonnieli commented on controllers/convoController.js:
```diff
@@ -104,10 +102,64 @@ const fetchAllMyConversationsGet = async (req, res) => {
+
+			// Filters in messages that need to be translated
+			recipientMessages.forEach((message) => {
+				if (!message.languages.includes(req.user.language)) {
```
you could also move this logic down so for example in worst case scenario, everything is not translated so it would be O(N), N being the total num of `recipientMessages`, and then in `filteredRecipientMessages`, you'll have to do another loop to actually translate it

you can just loop through `recipientMessages`, and combine it with your translation and check if it needs to be translated it from line 142 to 145

garzeah replied:
Thanks Bonnie, I'll implement that in a future PR!

garzeah committed:
Merge branch 'real-time-messaging' into google-translate-api
